netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+
+import numpy as np
+from functools import partial
+
+from netket.stats import sum_inplace, subtract_mean
+from netket.utils import n_nodes
+import netket.jax as nkjax
+
+from netket.utils.types import Array, Callable, PyTree, Scalar
+
+from .qgt_onthefly_logic import tree_cast, tree_conj, tree_axpy
```

Comment:
I wonder if we could move these functions into `netket/jax/utils.py` -- they are clearly more generally useful than just the onthefly logic!

Reply:
afaik i already moved them in #661

Reply:
Nice, I'll fix the imports then

Reply:
So the code is duplicated?

Reply:
Nope, but it's now imported from the wrong place apparently, should be `from netket.jax`

Reply:
so probably all tests are gonna fail...